src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,443 @@
+import java.util.Iterator;
+import java.util.Objects;
+
+import sun.security.util.Debug;
+import sun.security.jca.*;
+import sun.security.jca.GetInstance.Instance;
+
+import javax.crypto.spec.KDFParameterSpec;
+
+/**
+ * This class provides the functionality of a key derivation algorithm for the Java Cryptographic
+ * Extension (JCE) framework.
+ * <p>
+ * {@code KeyDerivation} objects will be instantiated through the {@code getInstance} family of
+ * methods.  Key derivation algorithm names will follow a naming convention of
+ * <I>algorithm</I>/<I>PRF</I>.  The algorithm field will be the KDF name
+ * (e.g. HKDF, TLS-PRF, PBKDF2, etc.), while the PRF specifier identifies the underlying
+ * pseudorandom function (e.g. HmacSHA256).  For instance, a KDF implementation of HKDF using
+ * HMAC-SHA256 will have an algorithm string of "HKDF/HmacSHA256".  In some cases the PRF portion of
+ * the algorithm specifier may be omitted if the KDF algorithm has a fixed or default PRF.
+ * <p>
+ * TODO: finish this javadoc
+ */
+
+public final class KDF {
+    private static final Debug debug = Debug.getInstance("jca", "KeyDerivation");
+
+    private static final Debug pdebug = Debug.getInstance("provider", "Provider");
+    private static final boolean skipDebug = Debug.isOn("engine=") && !Debug.isOn("keyderive");
+
+    // The provider
+    private Provider provider;
+
+    // The provider implementation (delegate)
+    private KDFSpi spi;
+
+    // The name of the MAC algorithm.
+    private final String algorithm;
+
+    // Additional KDF configuration parameters
+    private final AlgorithmParameterSpec algorithmParameterSpec;
+
+    // Has this object been initialized?
+    private boolean initialized = false;
+
+    // next service to try in provider selection
+    // null once provider is selected
+    private Service firstService;
+
+    // remaining services to try in provider selection
+    // null once provider is selected
+    private Iterator<Service> serviceIterator;
+
+    private final Object lock;
+
+    /**
+     * Instantiates a KeyDerivation object.
+     *
+     * @param keyDerivSpi
+     *     the delegate
+     * @param provider
+     *     the provider
+     * @param algorithm
+     *     the algorithm
+     * @param algParameterSpec
+     *     the algorithm parameters
+     */
+    protected KDF(KDFSpi keyDerivSpi, Provider provider, String algorithm,
+                  AlgorithmParameterSpec algParameterSpec) {
+        this.spi = keyDerivSpi;
+        this.provider = provider;
+        this.algorithm = algorithm;
+        this.algorithmParameterSpec = algParameterSpec;
+        this.lock = null;
+    }
+
+    /**
+     * Returns the algorithm name of this {@code KeyDerivation} object.
+     *
+     * <p>This is the same name that was specified in one of the
+     * {@code getInstance} calls that created this {@code KeyDerivation} object.
+     *
+     * @return the algorithm name of this {@code KeyDerivation} object.
```

Comment:
Nit: remove period, sentences that don't start with a capital letter and don't have a following sentence should not have period at end

Reply:
You missed a few in the `@throws` clauses.